sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -29,60 +30,76 @@ def _find_binning_thresholds(data, max_bins, subsample, random_state):
-        if len(distinct_values) <= max_bins:
-            midpoints = distinct_values[:-1] + distinct_values[1:]
-            midpoints *= .5
-        else:
-            # We sort again the data in this case. We could compute
-            # approximate midpoint percentiles using the output of
-            # np.unique(col_data, return_counts) instead but this is more
-            # work and the performance benefit will be limited because we
-            # work on a fixed-size subsample of the full data.
-            percentiles = np.linspace(0, 100, num=max_bins + 1)
-            percentiles = percentiles[1:-1]
-            midpoints = np.percentile(col_data, percentiles,
-                                      interpolation='midpoint').astype(X_DTYPE)
-            assert midpoints.shape[0] == max_bins - 1
+    # ignore missing values when computing bin thresholds
+    missing_mask = np.isnan(col_data)
+    if missing_mask.any():
+        col_data = col_data[~missing_mask]
+    col_data = np.ascontiguousarray(col_data, dtype=X_DTYPE)
+    distinct_values = np.unique(col_data)
+    if len(distinct_values) <= max_bins:
+        midpoints = distinct_values[:-1] + distinct_values[1:]
+        midpoints *= .5
+    else:
+        # We sort again the data in this case. We could compute
+        # approximate midpoint percentiles using the output of
+        # np.unique(col_data, return_counts) instead but this is more
+        # work and the performance benefit will be limited because we
+        # work on a fixed-size subsample of the full data.
+        percentiles = np.linspace(0, 100, num=max_bins + 1)
+        percentiles = percentiles[1:-1]
+        midpoints = np.percentile(col_data, percentiles,
+                                  interpolation='midpoint').astype(X_DTYPE)
+        assert midpoints.shape[0] == max_bins - 1
+
+    # We avoid having +inf thresholds: +inf thresholds are only allowed in
+    # a "split on nan" situation.
+    np.clip(midpoints, a_min=None, a_max=ALMOST_INF, out=midpoints)
+    return midpoints
+
+
+def _find_bin_categories(col_data, max_bins):
+    """Extract feature-wise categories from categorical data
+
+    Missing values and negative values are ignored. They will be considered
+    missing when ``_encode_categories`` is called.
+
+    Parameters
+    ----------
+    data : ndarray of shape (n_samples, n_features)
+        The data to bin.
+    max_bins: int
+        The maximum number of bins to be used for categories.
+
+    Return
+    ------
+    bin: ndarray
+        Map from bin index to categorical value. The size of each array is
+        equal to minimum of `max_bins` and categories' cardinality.
+    """
+    categories, counts = np.unique(col_data, return_counts=True)
 
-        # We avoid having +inf thresholds: +inf thresholds are only allowed in
-        # a "split on nan" situation.
-        np.clip(midpoints, a_min=None, a_max=ALMOST_INF, out=midpoints)
+    # sort by highest count
```

Comment:
Should we sort and re-order only when the cardinality exceeds max_bins? max_bins is small for now but if one day we allow for more than 256, it may matter.